11-bug_tracking.Rmd

```diff
@@ -1,23 +1,125 @@
+
+```{r, results='hide'}
+##  *** caught segfault ***
+## address (nil), cause 'memory not mapped'
+```
+
+2. If the code does not do what the documentation says it should, then either the code or the documentation is wrong. Report either of which needs to be fixed.
+
+$\textbf{Note}$:
+When you are in doubt that there is a bug:
+
+1. Make sure whether the bug appears in a clean session of $\textsf{R}$. Many a times, there are variables/commands/functions stored in history which might causes issues. Hence, check in a clean session. To do so, launch $\textsf{R}$ from the command line with the `--vanilla` option. 
+
+2. At times the code that is written is very complicated, has numerous package and file dependencies, has many function calls, etc.. In such scenarios it is quite common that the code throws an error and you are not able to solve it. You may tend to think that there is a bug that needs to be reported. Before doing so, try to produce a minimum working example of the code for the section where the error occurred. Add only those package and files which are required by that section, and see if the error appears still. Using this approach shall solve most of the errors.
+
+## What may not be a bug?
+
+1. In case the code is doing something unexpected, it may not necessarily be a bug. Carefully review the documentation of the function being called, and check whether the behaviour being exhibited on calling this function is the same as it was designed to do.
+
+2. Issues with $\textit{seemingly}$ identical numbers not being equal (especially floating point numbers) are usually not bugs. 
+
+3. If $\textsf{R}$ is running slower than expected, then also it may not be a bug. Ask someone else to review your code in such a case.
+
+4. If some function is working, but it is not defined in the best generalised way, then consult some to look over your code. This may perhaps not be a bug, instead it might be an alternative way of writing the function.
+
+## Checking if a bug is already reported
+
+The first step before filing a bug report is to see whether the problem has already been reported. Checking if the bug is reported will:
+
+1. Save time for you and the developers
+
+2. Help you see if the bug is already fixed for the next release
+
+3. Lead you to learn what needs to be done to fix it
+
+4. Determine if any additional information is needed
+
+The sections that follow discuss where to check whether a bug is already reported.
+
+## Levels of contributing to bug / What do you do when you find a bug?
+
+1. [Report](#ReportBug) the bug (if it is not alreasy reported).
+
+2. [Test](provide hyperlink to the section/chapter which discusses this) the bug.
+
+3. [Fix](hyperlink to the chapter on the lifecycle of a patch) the bug: Fixing a bug might require relatively more time. You may start a conversation about it either on BugZilla (link here) or on the GitHub mirror (link here). This would require engagement with $\textsf{R}$ Core team.

```

Comment:
Fixing bugs will be discussed in the chapter on lifecycle of a patch. How to test bugs will probably be a part of the next chapter on reviewing bugs. Let's see how those chapters shape up!

Reply:
Ok, let's see. I hope that the different chapters can be merged soon, it will help to have a more general read of the book and see if the sections match well. Is there any plan on when to merge #3 or this PR?

Reply:
I hope it is merged soon.